src/System.Private.CoreLib/shared/System/ReadOnlySpan.Fast.cs

```diff
@@ -154,6 +154,24 @@ internal ReadOnlySpan(ref T ptr, int length)
 #endif
         }
 
+        public T this[Index index]
+        {
+            get
+            {
+                return index.FromEnd ? this[_length - index.Value] : this[index.Value];
```

Comment:
If you can have `Span` reach in to the internals of `Index` and access the raw `_value` field you can probably improve the performance of this in perf-sensitive code paths.

Reply:
Another option is that if you allow `Index._value` to use a two's complement representation (rather than a one's complement representation), the getter here becomes simply:

```cs
return ref this[(_length & (index._rawValue >> 31)) + index._rawValue];
```

Reply:
>  to use a two's complement representation 

I assume the one's complement was picked intentionally to allow `(0, fromEnd: true)`. I do not think one's complement would work for it.

Reply:
I assume fromEnd is a literal and that the JIT can elide the branch during Index construction, but that the JIT cannot similarly reason about this indexer because the Index object construction may be far removed from this method call.

Reply:
> I assume fromEnd is a literal and that the JIT can elide the branch during Index construction, but that the JIT cannot similarly reason about this indexer because the Index object construction may be far removed from this method call.

That's right. I do not think it affects whether it is possible to use two's complement for the internal representation.

Reply:
> I assume the one's complement was picked intentionally to allow (0, fromEnd: true)

Correct. Specifically, `(0, fromEnd: true)` means "end of the sequence". That doesn't mean much for index, where `s[^0]` goes out of bounds, but means a lot for `s[2..^0]` which, since Range is exclusive-end, means every element starting from two and going to the end of the sequence.

Reply:
I'll keep the current implementation then.